### ISSUE TITLE
PORT is not showing for prometheus container while using docker-compose

### ISSUE THREAD

**Issue by eno-lc** [docker/compose#10497]
### Description

Hi guys,

I've created the following docker-compose file to create a prometheus container to consume an application stats, metrics and stuff. 
I am binding the ports that I want to use but when I run the docker-compose, the PORT is not shown on 'docker ps', and when I try to access it locally, I cannot because there isn't any PORT binded to the container.

Check out the docker-compose.yml file
![image](https://user-images.githubusercontent.com/93098680/234503172-917f0027-4598-4113-adcd-ba9acb99afa2.png)

Check out the console
![image](https://user-images.githubusercontent.com/93098680/234503455-70ef6aad-10a0-4935-943d-b4e0ddcd104f.png)

### Steps To Reproduce

1. Use my docker-compose file
2. Docker-compose it up
3. See the container - 'docker ps'

### Compose Version

```Text
Docker Compose version v2.10.2
```

### Docker Environment

```Text
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.9.1)
  compose: Docker Compose (Docker Inc., v2.10.2)
  extension: Manages Docker extensions (Docker Inc., v0.2.9)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.19.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 47
 Server Version: 20.10.17
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.16.3-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 6.696GiB
 Name: docker-desktop
 ID: 4SHX:2LHP:VMFE:RL5L:TZ6K:GWXK:X5JW:PUXV:JCPX:5TBT:FS3I:4C3F
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by glours** [comment]
Hello @eno-lc 
It seems you're using a old version of Compose (released in August 22), maybe also an old version of Docker Desktop.
Can you first try to update to the latest version to check if you can reproduce your issue?
Can you also share the result of a `docker inspect prometheus`, so we could check internal declaration of the container?

**Comment by eno-lc** [comment]
Hi @glours , sorry for the late reply did not get a notification.  I am trying the same on another machine that has Docker Compose version v2.17.3 , and Docker Desktop 4.19.0 (106363) is currently the newest version available. 

Results of docker inspect prometheus;
 docker inspect prometheus
[
    {
        "Id": "12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c",
        "Created": "2023-04-28T07:31:13.984578255Z",
        "Path": "/bin/prometheus",
        "Args": [
            "--config.file=/etc/prometheus/prometheus.yaml"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 2260,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2023-04-28T07:31:14.400637603Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:a5bac665ffa2c7c5a6fa96e9e9e00b51af36a5ca78cdbd682d8a90b6107aec84",
        "ResolvConfPath": "/var/lib/docker/containers/12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c/hostname",
        "HostsPath": "/var/lib/docker/containers/12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c/hosts",
        "LogPath": "/var/lib/docker/containers/12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c/12f21352f8f58cba8fadfcafc82fb790a0d748789e6b686a94307dcbace98f0c-json.log",
        "Name": "/prometheus",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "C:\\Users\\Sindroma\\Desktop\\grafana-prometheus-main\\data\\prometheus\\config:/etc/prometheus:rw"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "host",
            "PortBindings": {
                "9090/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "9090"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "ConsoleSize": [
                0,
                0
            ],
            "CapAdd": null,
            "CapDrop": null,
            "CgroupnsMode": "host",
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": [],
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/54051725db67aefdcfe256831119375533db5c9699829be9a69c34923659c42f-init/diff:/var/lib/docker/overlay2/38c5974cabba3662232521fdcfb1c02298093db4da6b8453ca8f34fe4810fe90/diff:/v
ar/lib/docker/overlay2/3ebecdd584a219bceef0c9037d1f25145edf68bcf2733314891010982d4dace4/diff:/var/lib/docker/overlay2/43f72a337a8b559d5e952b29114f7f796be8b0139749e460ae8852ab1f8f2864/diff:/var/lib/docker/overlay2/5fd5d8ab5934a
cb47edfaa70c95721542bcfccca4e2877af191e2c0b0e5f843d/diff:/var/lib/docker/overlay2/35a951fa22306fb3132bbd9cd83c1d31f20996a8987f7f22a259546a59a963c2/diff:/var/lib/docker/overlay2/9e811ac6bbd44049ebb96241481a0c2bdb562038fabb7f925
1753ae3d2c3faba/diff:/var/lib/docker/overlay2/cd3182606631de6da745fce10747cb7ec25b6641c7a9e9df6c59bde549fd0db7/diff:/var/lib/docker/overlay2/b6f341a9938ef2d25018ef8a2f0780af61e38de5882a04bd6b603a9d42e91903/diff:/var/lib/docker
/overlay2/c557c41f14c94086e6ddf034f9145eed6155768b775832535c23435c2ea2251e/diff:/var/lib/docker/overlay2/252e9e563f3223eb9e1500942e811b5807591829862e9f1d87ca3ea2d7e86ed1/diff:/var/lib/docker/overlay2/5da038f5452001071e9b7de376
0b6ffb41e509c9d961f632db0b7535a5e388ff/diff:/var/lib/docker/overlay2/19b99ba66b6e6a97af404a7f5fb5699954c19638b609bb2bf660b922fcab5c11/diff",
                "MergedDir": "/var/lib/docker/overlay2/54051725db67aefdcfe256831119375533db5c9699829be9a69c34923659c42f/merged",
                "UpperDir": "/var/lib/docker/overlay2/54051725db67aefdcfe256831119375533db5c9699829be9a69c34923659c42f/diff",
                "WorkDir": "/var/lib/docker/overlay2/54051725db67aefdcfe256831119375533db5c9699829be9a69c34923659c42f/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "C:\\Users\\Sindroma\\Desktop\\grafana-prometheus-main\\data\\prometheus\\config",
                "Destination": "/etc/prometheus",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "volume",
                "Name": "7f5d955e0fa43f8cd1e159f85157fc171fefc0db2a74760daa810d6450e8bc6d",
                "Source": "/var/lib/docker/volumes/7f5d955e0fa43f8cd1e159f85157fc171fefc0db2a74760daa810d6450e8bc6d/_data",
                "Destination": "/prometheus",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "docker-desktop",
            "Domainname": "",
            "User": "nobody",
            "AttachStdin": false,
            "AttachStdout": true,
            "AttachStderr": true,
            "ExposedPorts": {
                "9090/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "--config.file=/etc/prometheus/prometheus.yaml"
            ],
            "Image": "prom/prometheus:v2.35.0",
            "Volumes": {
                "/prometheus": {}
            },
            "WorkingDir": "/prometheus",
            "Entrypoint": [
                "/bin/prometheus"
            ],
            "OnBuild": null,
            "Labels": {
                "com.docker.compose.config-hash": "2e02fa5ea67f0adc4ed53f8f3c3e34e8d1701792ce8385e74181153a2c02396f",
                "com.docker.compose.container-number": "1",
                "com.docker.compose.depends_on": "",
                "com.docker.compose.image": "sha256:a5bac665ffa2c7c5a6fa96e9e9e00b51af36a5ca78cdbd682d8a90b6107aec84",
                "com.docker.compose.oneoff": "False",
                "com.docker.compose.project": "grafana-prometheus-main",
                "com.docker.compose.project.config_files": "C:\\Users\\Sindroma\\Desktop\\grafana-prometheus-main\\docker-compose.yaml",
                "com.docker.compose.project.working_dir": "C:\\Users\\Sindroma\\Desktop\\grafana-prometheus-main",
                "com.docker.compose.service": "prometheus",
                "com.docker.compose.version": "2.17.3",
                "maintainer": "The Prometheus Authors \u003cprometheus-developers@googlegroups.com\u003e"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "6dca7c93fb73c2ddf3487d2231ce61c8f46c8c9edeb04f0d4d0ac34fa24e1698",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "",
                    "DriverOpts": null
                }
            }
        }
    }
]

**Comment by glours** [comment]
And did you test the output of `docker ps` with those versions of Docker Desktop and Compose? 
Did you notice the same result?

**Comment by milas** [comment]
Port mappings are ignored with `network_mode: host`.

The Docker CLI warns about this, we could consider doing the same in Compose:
```
❯ docker run --rm -it --network=host -p 80 nginx
WARNING: Published ports are discarded when using host network mode
```

When using host-mode networking, the container shares the network interface with the host, so any ports it listens on are already available. In general, host-mode networking is only recommended if required for a specific reason (e.g. UDP multicast traffic).

Additionally, as you're using Docker Desktop, host-mode networking will not behave as expected: the "host" networking device is actually the Docker Desktop-managed Linux VM.

---

tl;dr remove `network_mode: host` from your config and re-launch

**Comment by eno-lc** [comment]
> Port mappings are ignored with `network_mode: host`.
> 
> The Docker CLI warns about this, we could consider doing the same in Compose:
> 
> ```
> ❯ docker run --rm -it --network=host -p 80 nginx
> WARNING: Published ports are discarded when using host network mode
> ```
> 
> When using host-mode networking, the container shares the network interface with the host, so any ports it listens on are already available. In general, host-mode networking is only recommended if required for a specific reason (e.g. UDP multicast traffic).
> 
> Additionally, as you're using Docker Desktop, host-mode networking will not behave as expected: the "host" networking device is actually the Docker Desktop-managed Linux VM.
> 
> tl;dr remove `network_mode: host` from your config and re-launch

The reason why I was using host-mode networking is because I had a Spring Application running locally as a jar file, and had Grafana and Prometheus running as Docker containers, I know it would work if I created the Spring Application as a Docker container and then created a common network for the 3 containers, but just wanted to check how it works with host-mode networking, might as well just make them all as Docker Containers and create a common network for them to communicate. Either way thank you for the clarification. Have a good one! :+1:  :)

**Comment by Mitko101513** [comment]
@eno-lc our best shot is to create docker container for the Spring Application and use "bridge" network_mode for the whole stack?

**Comment by milas** [comment]
@Mitko101513 Take a look at the [Networking in Compose](https://docs.docker.com/compose/networking/) guide.

By default, Compose creates a network for all the services in your project. They'll be able to access any ports on each other as well as have DNS resolution by service name. From the host, nothing will be accessible by default, and you can use port mappings to open what you need from the host.

So, for example, you could run Grafana + Prometheus via Compose, and add port mappings for them, then run the Spring app on your host. For example, let's say you have `1234:9090` on the Prometheus port. An app running on the host could access that by `localhost:1234`. An app in a different service in the Compose project could access that by `prometheus:9090` (regardless of whether the port mapping is created or what port it maps to on the host).

**Comment by Mitko101513** [comment]
Thank you @milas for the clarification, I really appreciate it

**Comment by pjmenezes** [comment]
I encountered the same issue on MacOS 14.2.1. To resolve it, I had to create a network and attach a new subnet to it. The **--subnet** option is used to specify the range of the custom subnet, and **br0** is the bridge.
Here is the solution that worked for me:
`docker network create --subnet=192.168.1.0/24 --driver=bridge my-network br0
`

Check the documentation:[ https://docs.docker.com/engine/reference/commandline/network_create/](https://docs.docker.com/engine/reference/commandline/network_create/)
Additionally, you can assign the IP range of the subnet to the network using the **--ip-range** option and the **--gateway** option to set the network gateway. I found this practical guide:
[https://webdock.io/en/docs/how-guides/docker-guides/how-to-create-and-manage-docker-networks-and-docker-volumes](https://webdock.io/en/docs/how-guides/docker-guides/how-to-create-and-manage-docker-networks-and-docker-volumes)  
Example presented in the guid: 
`docker network create --driver=bridge --subnet=10.0.0.0/16 --ip-range=10.0.0.0/24 --gateway=10.0.0.1 eth2`
